### DISCUSSION TITLE
SQL LIKE/pattern matching

### DISCUSSION THREAD

**Discussion by mkoo21** [prisma/prisma#3159] [Q&A] [answered]
In SQL we can do things like `SELECT * FROM Product WHERE Title Like '%bicycle%light%'` (suppose I'm looking for a nice set of headlights for my bike at night). The docs seem to indicate that the closest thing we can do for filtering in prisma to this is `contains`, which doesn't have wildcards - so is the only way to achieve something like this to use a raw query?

**Comment by ryands17** [comment, marked as answer]
Hey @mkoo21 👋 
Unfortunately, the best way to do this would be using `prisma.queryRaw` as pattern matching is not yet supported in the API. Do add a 👍 on [this](https://github.com/prisma/prisma/issues/1684) feature request so that we can take a look at the priority of this.

**Comment by Epskampie** [comment]
Here is a full example using `queryRaw`:

```js
const query = `%${'SEARCH_QUERY_HERE'}%`;
const maxResults = 10;

// Have to use queryRaw because "LIKE" is not supported in prisma yet
const ids = await prisma.$queryRaw<{ id: number }[]>(`
  SELECT id FROM "User"
  WHERE "firstName" LIKE '${query}'
  LIMIT ${maxResults}
  ;`);

const users = await prisma.user.findMany({
  where: { id: { in: ids.map((row) => row.id) } },
});
```

**Reply by keslert** [reply]
I had to drop the single quotes to make it work

```
const tags = await prisma.$queryRaw<Tag[]>`
  SELECT * FROM Tag 
  WHERE slug LIKE ${`%${slug}%`}
  ORDER BY LENGTH(slug)
  LIMIT 10
;`
```

**Reply by neicore** [reply]
Thank you for this comment. saved my life

**Reply by Joshua-Hager** [reply]
Thank you! Found this after struggling with different versions of string interpolation trying to get it to work.

**Reply by jstkiddin** [reply]
I've been straggling with this one for a long time. Thank you a lot!

**Reply by Varkoff** [reply]
Thanks ! I've been facing the same problem !

**Reply by hmake98** [reply]
Thanks for sharing

**Reply by nas** [reply]
Any idea when prisma is adding a regex query feature e.g. 
```
SELECT * FROM your_table WHERE your_column ~ '^[A-Z]{3}-\d+$'; ?
```

**Comment by Epskampie** [comment]
You can use `contains` / `startsWith` to have something similar to `LIKE` for simple cases:

https://www.prisma.io/docs/concepts/components/prisma-client/filtering-and-sorting

```
const res = await prisma.post.findMany({
  where: {
    author: {
      email: {
        contains: 'prisma.io',
      },
    },
  },
})
```

**Reply by ctilley83** [reply]
You can't do this with dates though which can be accomplished using LIKE.

**Comment by shimaxu** [comment]
They have preview feature for this
https://www.prisma.io/docs/concepts/components/prisma-client/full-text-search

**Comment by mkoo21** [comment]
Thank you for the random replies over the past 2 years but it seems the _correct_ answer to the question is still @ryands17 original response (ie. no, use the `rawQuery` escape hatch). A "bicycle light" is not that the same as a "light bicycle" and this "full text search" is still a weak alternative to pattern matching.

**Reply by binajmen** [reply]
> Thank you for the random replies over the past 2 years but it seems the _correct_ answer to the question is still @ryands17 original response (ie. no, use the `rawQuery` escape hatch). A "bicycle light" is not that the same as a "light bicycle" and this "full text search" is still a weak alternative to pattern matching.

Can you expand why you think this is a weak alternative? :curious:

**Comment by todordev** [comment]
Can I search for a string with '_contains_' if the string includes _ or %?
I try to search for **_en** but Prisma does not take that symbol.
It returns results that contain 'en' not '_en'.

What can I do?

```javascript
const res = await prisma.post.findMany({
  where: {
    name: {
      contains: '_en',
    },
  },
})
```

Thank you in advance!

**Reply by mcjambi** [reply]
i have the same question, and prisma sometime return a very wrong value, startsWith and contains return null at all, sometime not correct result

**Reply by janpio** [reply]
If Prisma returns invalid or wrong data, please open an issue and provide all the information the issue template asks for. We are _very_ interested in fixing these. Thanks.

**Reply by mcjambi** [reply]
actually, i can not wait for fixing an issue, i have used my custom query  to search and match keyword, prisma doesn't do it correctly...

**Reply by databrecht** [reply]
You might be able to derive it from the query it translates to in your underlying database and then search for the escape character that could be used. 

In postgres (my case) it translates to a 'Like', if I look for escape chars: "The escape-character is an optional parameter. It allows for testing of literal instances of wildcard characters such as % and _. If it is not provided, the \ will be used as the escape character."

**Comment by databrecht** [comment]
To be fair towards the devs of Prisma and to be clear for other people who are searching for this I would like to add the following results of my tests (which is for the Postgres case, the query compilation will, of course, be different for other dbs) of how it behaves today since I assumed that things wouldn't work as I expect when reading this, while it seems it does work for my usecase.

**contains:** if you are using Postgres and looking for a like, use **contains** since that's what --at the time of writing-- it seems to translate to. It seems I'm able to just pass wildcards in the contains string so that seems to be (at least in Postgres) to do the same as the raw query that was proposed in the answer. 

**full-text preview**, this might not be what people are looking for when they are looking for 'like'. What it does (in Postgres) is transform to tsvector (https://www.postgresql.org/docs/current/datatype-textsearch.html), e.g. transform to lexemes used for true full-text searching (searching in big documents). If you would use this, let's say for searching on a field that contains a UUID, you are probably not going to get what you intended.  Definitely useful, different usecase and not a solution to what the original author was asking.

**Reply by mpatrickaires** [reply]
Thank you very much! 
Coming from 2024, this also worked for me using SQLite.

For example, the following would return all records where the field `name` contains a "John" anywhere on it, followed by a "Doe" anywhere after "John", whilst also being case-insensitive.

```typescript
where: {
  name: {
    contains: "%John%Doe%",
  },
},
```

**Comment by samirant15** [comment]
I got it working with the `prisma.$queryRaw` like this: ```WHERE name LIKE ${`%${search || ''}%`}```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
